src/vs/editor/common/model/textModelEvents.ts

```diff
@@ -126,6 +128,69 @@ export class ModelRawFlush {
 	public readonly changeType = RawContentChangedType.Flush;
 }
 
+/**
+ * Represents text injected on a line
+ * @internal
+ */
+export class LineInjectedText {
+	public static applyInjectedText(lineText: string, injectedTexts: LineInjectedText[] | null): string {
+		if (!injectedTexts || injectedTexts.length === 0) {
+			return lineText;
+		}
+		let result = '';
+		let lastOriginalOffset = 0;
+		for (const injectedText of injectedTexts) {
+			result += lineText.substring(lastOriginalOffset, injectedText.column - 1);
+			lastOriginalOffset = injectedText.column - 1;
+			result += injectedText.text;
+		}
+		result += lineText.substring(lastOriginalOffset);
+		return result;
+	}
+
+	public static fromDecorations(decorations: IModelDecoration[]): LineInjectedText[] {
+		const result: LineInjectedText[] = [];
+		for (const decoration of decorations) {
+			if (decoration.options.before) {
```

Comment:
We should only include here decorations with `decoration.options.before.content.length > 0`. Having injected text with empty injected content might break some assumptions from `LineBreakData` / `monospaceLineBreaksComputer.ts` / etc.